### PR TITLE
Added Ordering instance to HttpDate class

### PR DESCRIPTION
**Pull request by san-coding**
Contributing to #5017 

Added `Ordering` instance to `HttpDate` ,  in branch `series/0.22`

### CODE REVIEW

**Comment by san-coding** [comment]
@rossabaker , have added `Ordering` instance to `HttpDate` , as expected the format tests are failing

**Comment by san-coding** [comment]
@rossabaker , compile checks are also failing, I did what you suggested, 
```
  implicit val stdLibOrderingInstance: Ordering[HttpDate] =  Order[HttpDate].toOrdering
```
But it is showing error

**Comment by san-coding** [comment]
@rossabaker I added import `cats.kernel.Order` , but now i am getting the following error :

> /home/runner/work/http4s/http4s/dsl/src/main/scala-3/org/http4s/dsl/impl/Auth.scala:23: error: ; expected but def found
>     infix def unapply[F[_], A](ar: AuthedRequest[F, A]): Option[(Request[F], A)] =
>           ^
> [error] -- Error: /home/runner/work/http4s/http4s/core/src/main/scala/org/http4s/HttpDate.scala:58:75 
> [error] 58 |  implicit val stdLibOrderingInstance: Ordering[HttpDate] = Order[HttpDate].toOrdering
> [error]    |                                                                           ^
> [error]    |no implicit argument of type cats.kernel.Order[org.http4s.HttpDate] was found for parameter ev of method apply in object Order
> [error] one error found
> [error] one error found
> [error] (core / Compile / compileIncremental) Compilation failed

**Comment by san-coding** [comment]
I made the change but once again the formatting checks are failing

**Comment by rossabaker** [comment]
My mistake.  I thought we had a `cats.kernel.Order` instance for this.  That's a lot more interesting than the standard library one, which can be implicitly derived from the Cats one.  Let's try this.  Do you have a working compiler yet?

Add this to `HttpDate` instead of what you have:

```scala
  implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
    new Order[HttpDate] {

    }
```

You'll need to add an import of `cats.Order`.  The compiler will tell you what method is missing.  See if you can add it and implemented.  When that works, we can discuss testing.

**Comment by san-coding** [comment]
> My mistake. I thought we had a `cats.kernel.Order` instance for this. That's a lot more interesting than the standard library one, which can be implicitly derived from the Cats one. Let's try this. Do you have a working compiler yet?
> 
> Add this to `HttpDate` instead of what you have:
> 
> ```scala
>   implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
>     new Order[HttpDate] {
> 
>     }
> ```
> 
> You'll need to add an import of `cats.Order`. The compiler will tell you what method is missing. See if you can add it and implemented. When that works, we can discuss testing.

Alright thanks @rossabaker , I will look into it

**Comment by rossabaker** [comment]
Yes, what he suggested should work.  I got you in trouble because I mistook that we already had the Cats instance.

**Comment by san-coding** [comment]
Hey @rossabaker , I have also overridden compare method, is that needed? what all should inside the catsOrderForHttp4sHttpDate instance {}

**Comment by san-coding** [comment]
Looks like Compile check are failing because  

> value headerName is not a member of org.http4s.HttpDate

**Comment by san-coding** [comment]
So, what should I put inside instance, usually I have seen the compare function, but I added compare and it gave the above error
> 
> ```scala
>   implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
>     new Order[HttpDate] {
> 
>     }
> ```
>

File: core/src/main/scala/org/http4s/HttpDate.scala [hunk 1]
```diff
@@ -54,6 +55,11 @@ class HttpDate private (val epochSecond: Long) extends Renderable with Ordered[H
 object HttpDate {
   private val MinEpochSecond = -2208988800L
   private val MaxEpochSecond = 253402300799L
+  implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
```
Comment (isomarcte) [inline core/src/main/scala/org/http4s/HttpDate.scala:58]: @san-coding if you are going to add add an `Order` can you also add a `Hash` with it? Something like we do here https://github.com/http4s/http4s/blob/main/core/shared/src/main/scala/org/http4s/RequestPrelude.scala#L102

Sorry, this is extra confusing and weird. The reason we have to do this is that both [Order](https://github.com/http4s/http4s/blob/main/core/shared/src/main/scala/org/http4s/RequestPrelude.scala#L102) and [Hash](https://github.com/typelevel/cats/blob/e0784fc0d95b9c4eed0bb1ddce6c69e3fd9a4a4f/kernel/src/main/scala/cats/kernel/Hash.scala#L12) extend [Eq](https://github.com/typelevel/cats/blob/e0784fc0d95b9c4eed0bb1ddce6c69e3fd9a4a4f/kernel/src/main/scala/cats/kernel/Eq.scala#L18) and if they aren't both in the concrete instance we get into an ambiguous implicit context.

So, if we don't add a `Hash` instance at the same time we add an `Order` we'll have to do a deprecation in the future if we ever want to add a `Hash`.

I hope that makes sense....I'm sorry, it's very confusing.

**Review by isomarcte — CHANGES_REQUESTED** [review]

**Comment by san-coding** [comment]
Hey @rossabaker , `QValue` class is extending `Ordered`, and has a `cats.kernel` instance, but and `Ordering` instance has to be added to it right?
https://github.com/http4s/http4s/blob/a942a83e2b88bafc644fc9916fd1979eb951ced8/core/shared/src/main/scala/org/http4s/QValue.scala#L131

File: core/src/main/scala/org/http4s/HttpDate.scala (hunk 1 above)
Comment (san-coding) [reply core/src/main/scala/org/http4s/HttpDate.scala:58]: Ok I will add Hash, but can you tell me what I should put inside the instance definition 
```
 implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
    new Order[HttpDate] {
    What should add here
    }
```

**Comment by san-coding** [comment]
@rossabaker @isomarcte 
When I had added `def compare()` definition inside the instance :
```
  implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
    new Order[HttpDate] {
      override def compare(x: HttpDate, y: HttpDate): Int =
        x.headerName.compareTo(y.headerName)
    }
```
It gave this error : 
> value headerName is not a member of org.http4s.HttpDate

Then I removed the `def compare()` and left it empty like this

```
  implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
    new Order[HttpDate] {
    }
```
Now it says `object` cannot be created because `def compare()` is missing

> [error] -- Error: /home/runner/work/http4s/http4s/core/src/main/scala/org/http4s/HttpDate.scala:59:4 
> 31
> [error] 59 |    new Order[HttpDate] {
> 32
> [error]    |    ^
> 33
> [error]    |object creation impossible, since def compare(x: A, y: A): Int in trait Order in package cats.kernel is not defined

**Comment by san-coding** [comment]
Hurray 🎉, finally it is compiling , so this is the cats instance definition I added 
```
  implicit val catsOrderForHttp4sHttpDate: Order[HttpDate] =
    new Order[HttpDate] {
      override def compare(x: HttpDate, y: HttpDate): Int =
        x.headerName.compareTo(y.headerName)
    }
```
@isomarcte @rossabaker , only formatting checks are failing ,I will add `Hash` soon, please review and give your feedback

**Comment by san-coding** [comment]
Hey @isomarcte @rossabaker , I have made the changes you requested:  added `Hash` ,compile checks have passed , only formatting checks are failing, I think  PR is ready for review once someone helps out with `sbt scalafmtAll` commit

**Comment by rossabaker** [comment]
Do you have sbt working yet?  Or Metals?  Either of those should be able to do the formatting for you.

I took care of this one.

**Comment by san-coding** [comment]
Thanks @rossabaker @isomarcte , should I make anymore changes or is this PR ready to be merged

**Comment by san-coding** [comment]
Which classes should I next add `Ordering` instance

**Comment by san-coding** [comment]
> Do you have sbt working yet? Or Metals? Either of those should be able to do the formatting for you.
> 
> I took care of this one.

I have installed metals, but set doesn't seem to be working

**Comment by san-coding** [comment]
I feel that scastie should work if given the right configuration, can anyone help me out with what configurations are to be given
image

**Comment by san-coding** [comment]
> Do you have sbt working yet? Or Metals? Either of those should be able to do the formatting for you.
> 
> I took care of this one.

I think I have figured it out on VSCode 🎉

**Comment by rossabaker** [comment]
This should have tests, too.  Look at `HttpVersionSuite` for how it has `checkAll` on `OrderTests` and `HashTests`.  That's how we know that the instances you created are lawful.

**Comment by san-coding** [comment]
> This should have tests, too. Look at `HttpVersionSuite` for how it has `checkAll` on `OrderTests` and `HashTests`. That's how we know that the instances you created are lawful.

Ohh okay, I will make the required changes @rossabaker , thanks

**Comment by san-coding** [comment]
> This should have tests, too. Look at `HttpVersionSuite` for how it has `checkAll` on `OrderTests` and `HashTests`. That's how we know that the instances you created are lawful.

You mean I have to add a `OrderingTests` , should I make a `HttpDateSuite.scala` ? I am a little confused here, @isomarcte @rossabaker

**Comment by rossabaker** [comment]
Adding tests [like this](https://github.com/http4s/http4s/blob/0ca03858cdfc5aedc0966a7e92f928540083a1e1/tests/shared/src/test/scala/org/http4s/HttpVersionSuite.scala#L27-L29) to the existing `HttpDateSuite` on this branch should get the job done.  You will probably need to add imports: see the linked file for an example.

**Comment by san-coding** [comment]
> Adding tests [like this](https://github.com/http4s/http4s/blob/0ca03858cdfc5aedc0966a7e92f928540083a1e1/tests/shared/src/test/scala/org/http4s/HttpVersionSuite.scala#L27-L29) to the existing `HttpDateSuite` on this branch should get the job done. You will probably need to add imports: see the linked file for an example.

Ok , I will add the HttpDateSuite.scala to branch `series/0.22`

**Comment by san-coding** [comment]
@rossabaker I have added test and made the requested changes

**Review by rossabaker — APPROVED** [review]
:+1: if green

**Comment by san-coding** [comment]
@rossabaker , all the tests have passed, should I make any more changes in this PR ? Thanks

**Comment by rossabaker** [comment]
if you can search for files that have `Ordered` and don't have a Cats `Order`, that's great: the Cats instances are what are the most interesting.

**Comment by san-coding** [comment]
> if you can search for files that have `Ordered` and don't have a Cats `Order`, that's great: the Cats instances are what are the most interesting.

Thanks , will start working on it immediately 🎉

**Pull request merged**